src/commons/util/file.rs

```diff
@@ -46,81 +55,95 @@ pub fn file_path(base_path: &Path, file_name: &str) -> PathBuf {
     let mut full_path = base_path.to_path_buf();
     full_path.push(name);
     save(content, &full_path)
 }
 
 /// Saves a file under a base directory, using the rsync uri to create
 /// sub-directories preserving the rsync authority and module in dir names.
-pub fn save_with_rsync_uri(content: &Bytes, base_path: &Path, uri: &uri::Rsync) -> Result<(), io::Error> {
+pub fn save_with_rsync_uri(content: &Bytes, base_path: &Path, uri: &uri::Rsync) -> Result<(), KrillIoError> {
     let path = path_with_rsync(base_path, uri);
     save(content, &path)
 }
 
 /// Reads a file to Bytes
-pub fn read(path: &Path) -> Result<Bytes, io::Error> {
-    let mut f = File::open(path).map_err(|_| Error::cannot_read(path))?;
+pub fn read(path: &Path) -> Result<Bytes, KrillIoError> {
+    let mut f =
+        File::open(path).map_err(|e| KrillIoError::new(format!("Could not read: '{}'", path.to_string_lossy()), e))?;
     let mut bytes = Vec::new();
-    f.read_to_end(&mut bytes)?;
+    f.read_to_end(&mut bytes)
+        .map_err(|e| KrillIoError::new(format!("Could not read: {}", path.to_string_lossy()), e))?;
     Ok(Bytes::from(bytes))
 }
 
-pub fn read_with_rsync_uri(base_path: &Path, uri: &uri::Rsync) -> Result<Bytes, io::Error> {
+pub fn read_with_rsync_uri(base_path: &Path, uri: &uri::Rsync) -> Result<Bytes, KrillIoError> {
     let path = path_with_rsync(base_path, uri);
     read(&path)
 }
 
-pub fn delete_with_rsync_uri(base_path: &Path, uri: &uri::Rsync) -> Result<(), io::Error> {
+pub fn delete_with_rsync_uri(base_path: &Path, uri: &uri::Rsync) -> Result<(), KrillIoError> {
     delete(&path_with_rsync(base_path, uri))
 }
 
-pub fn delete_in_dir(base_path: &Path, name: &str) -> Result<(), io::Error> {
+pub fn delete_in_dir(base_path: &Path, name: &str) -> Result<(), KrillIoError> {
     let mut full_path = base_path.to_path_buf();
     full_path.push(name);
     delete(&full_path)
 }
 
-pub fn delete(full_path: &Path) -> Result<(), io::Error> {
+pub fn delete(full_path: &Path) -> Result<(), KrillIoError> {
     trace!("Removing file: {}", full_path.to_string_lossy());
-    fs::remove_file(full_path)?;
-    Ok(())
+    fs::remove_file(full_path)
+        .map_err(|e| KrillIoError::new(format!("Could not remove file: {}", full_path.to_string_lossy()), e))
 }
 
-pub fn clean_file_and_path(path: &Path) -> Result<(), io::Error> {
+pub fn clean_file_and_path(path: &Path) -> Result<(), KrillIoError> {
     if path.exists() {
-        fs::remove_file(&path)?;
+        fs::remove_file(&path)
+            .map_err(|e| KrillIoError::new(format!("Could not remove file: {}", path.to_string_lossy()), e))?;
 
         let mut parent_opt = path.parent();
 
         while parent_opt.is_some() {
             let parent = parent_opt.unwrap();
-            if parent.read_dir()?.count() == 0 {
+            if parent
+                .read_dir()
+                .map_err(|e| KrillIoError::new(format!("Could not read directory: '{}'", parent.to_string_lossy()), e))?
```

Comment:
The same error message is used in a different function above, will it be possible to tell which of the two cases occurred or should more context be included in these messages?

Reply:
I don't think it matters. There is context, it lists the directory and the error. This case is also extremely unlikely. This would be failure to read a parent dir after successfully removing a file under this dir (potentially some levels deeper)